### PR TITLE
fix(mac): increase OSK character size by 50%

### LINKED ISSUE #5600: feat(mac): Make Hebrew Consonants and Vowels Show Bigger in On-Screen Keyboard
**Is your feature request related to a problem? Please describe.**

The Hebrew consonants and the vowels in the On-Screen keyboard are too small, the vowels especially. Screenshot:
image

**Describe the solution you'd like**

Please make them bigger if possible.

**Describe alternatives you've considered**

Well, because it is almost unusable, I had to ask the students to print the mapping out.

---

**Keyman for macOS:**
 - OS:  macOS Big Sur
 - Keyman version:  13.0.101
 - Target Application:  Keyman OSK

---

**Keyboard**
 - Keyboard name:  sil_hebrew
 - Keyboard version:  1.8.1
 - Language name: Hebrew (SIL)

### PR DESCRIPTION
**Pull request by sgschantz**
Increases the size of character labels in the Keyman for Mac onscreen keyboard by 50% (from using 0.3 of the key height to 0.45 of the key height). The font size for labels on modifier and special keys remains as is. Fixes #5600 

OSK Hebrew 0 5 600
OSK Hebrew 03 600

# User Testing
* TEST_MAC: load keyboards for a variety of languages and view the OSK for each to verify that characters fit cleanly within the bounds of the key

### CODE REVIEW

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m [hunk 1]
```diff
@@ -141,6 +144,26 @@ - (void)setKeyCode:(NSUInteger)keyCode {
...
+        _isCharacterKey = YES;
+    else
+        _isCharacterKey = NO;
+    
+    // having changed the keycode, adjust the font size of the key label to optimize for available space
+    [self adjustLabelFontSize];
+}
+
+// set the size of the font relative to the key height, depending on what type of key it is
+// this allows larger more readable character key text
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m:158]:
```suggestion
/**
 * set the size of the font relative to the key height, depending on what type of key it is
 * this allows larger more readable character key text
 */
```

Per our [style guide](https://github.com/keymanapp/keyman/wiki/Keyman-Code-Style-Guide#Source-Code-Documentation-Syntax), we are trying to use javadoc-style comments for function headers.

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m [hunk 2]
```diff
@@ -141,6 +144,26 @@ - (void)setKeyCode:(NSUInteger)keyCode {
         _isSpecialKey = YES;
     else
         _isSpecialKey = NO;
+    
+    if (!_isSpecialKey && !_isModifierKey)
+        _isCharacterKey = YES;
+    else
+        _isCharacterKey = NO;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m:151]: For new code, use the [style guide](https://github.com/keymanapp/keyman/wiki/Keyman-Code-Style-Guide#code-format) for blocks (although spacing for new code should match file spacing):

```suggestion
    if (!_isSpecialKey && !_isModifierKey) {
        _isCharacterKey = YES;
    } else {
        _isCharacterKey = NO;
    }
```

(If XCode wants `else {` on a new line, we can live with that :wink:)

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m [hunk 3]
```diff
@@ -141,6 +144,26 @@ - (void)setKeyCode:(NSUInteger)keyCode {
...
+    [self adjustLabelFontSize];
+}
+
+// set the size of the font relative to the key height, depending on what type of key it is
+// this allows larger more readable character key text
+- (void)adjustLabelFontSize {
+    CGFloat relativeFontSize = 0.45;
+    
+    if (!self.isCharacterKey)
+        relativeFontSize = 0.3;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m:163]:
```suggestion
    if (!self.isCharacterKey) {
        relativeFontSize = 0.3;
    }
```

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m [hunk 4]
```diff
@@ -115,7 +118,7 @@ - (BOOL)isFlipped {
 }
 
 - (void)setKey:(OSKKey *)key {
-    _key = key;
+   _key = key;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m:121]:
```suggestion
    _key = key;
```

restore indent

File: mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m [hunk 5]
```diff
@@ -49,9 +49,12 @@ - (id)initWithFrame:(NSRect)frame {
         if ([_caption respondsToSelector:@selector(setLineBreakMode:)]) {
             [_label setLineBreakMode:NSLineBreakByClipping];
         } // There might be some problem not calling this, but it seems to be okay as far as I can tell
+        
+        CGFloat fontSize = labelFrame.size.height*0.3;
```
Comment (mcdurdin) [inline mac/KeymanEngine4Mac/KeymanEngine4Mac/KME/OnScreenKeyboard/KeyView.m:53]: I wonder if we can make the `0.3` and `0.45` values constants at the top of this file, especially given we use `0.3` in two places?

**Review by mcdurdin — APPROVED** [review]
LGTM

I only have nit picks in my comments 😁

**Comment by MakaraSok** [comment]
* **TEST_MAC: FAILED** Two issues found:
1. Text on some keys are cropped off, i.e.<kbd> េះ</kbd> and <kbd>ោះ</kbd>

   Screen Shot 2021-12-09 at 10 46 28 AM

2. When pressing on the OSK directly, the output of the Shift layer is like that of the default layer, i.e. press <kbd>Shift</kbd> and then <kbd>អ</kbd> to get the unexpected output of ង​ instead. This issue doesn't occur when typing with the physical keyboard.

    https://user-images.githubusercontent.com/28331388/145332319-d9d56c2b-4af4-438c-a36a-91cc2292aea1.mov

**Comment by sgschantz** [comment]
1. I don't see the first issue, but this could be device-dependent because of how the font size is calculated. I'll try making it a little smaller. One question: do the clipped key labels improve after you resize the OSK window?

2. I was able to reproduce this problem in other builds, so a new issue should be created for this. For me, the shift key does not work for the physical keyboard - I always get the default key mapping regardless of what modifier keys I hold down.

**Comment by sgschantz** [comment]
Adjusted the key label font size down slightly -- see how this looks
@keymanapp-test-bot retest [TEST_MAC]

**Comment by mcdurdin** [comment]
> Adjusted the key label font size down slightly -- see how this looks

I wonder if this is going to be a good answer -- I think it's always going to depend on the keyboard. Three longer term solutions:

1. Fix the double dotted circle issue for Khmer on the OSK
2. Have per-key scaling of text as required.
3. Ability to specify on screen keyboard font scaling per platform in packages (as specified by keyboard dev). (This requires the ability to preview the solution which will be a challenge!)

I think all of these are necessary long-term, and I like the larger font that we already have despite the fact that it isn't a 100% solution.

**Comment by MakaraSok** [comment]
> 1. I don't see the first issue, but this could be device-dependent because of how the font size is calculated. I'll try making it a little smaller. One question: do the clipped key labels improve after you resize the OSK window?

Resize the OSK doesn't help to fully show the text on one of the two keys in question.
image

I haven't tried the new build yet.

**Comment by MakaraSok** [comment]
* **TEST_MAC: FAILED** The text of <kbd>ោះ</kbd> on shift layer is still not fully visible.

Minimum size: the key is fully visible
image

Maximum size: the key is not fully visible, ះ​ the two circles are buried and covered under the key boundary. 
image

**Comment by MakaraSok** [comment]
@mcdurdin If the dotted circle is fixed, this might no longer be an issue. What happens now is that the two characters are spreader out instead of collating together. 

Second thought, can we do something like what we have in KeymanWeb. The key looks fully visible at all resizes.
image

**Comment by mcdurdin** [comment]
Can we restore the height here to 0.45 and create an issue to track per-key scaling for 16.0?

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.171-alpha](https://keyman.com/downloads/releases/alpha/15.0.171)

**Pull request merged**